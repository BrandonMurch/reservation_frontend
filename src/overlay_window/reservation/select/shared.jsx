// Dependencies
import React from 'react';

// eslint-disable-next-line
export const emptyOption = <option key={0} aria-label="emptyOption" value="" />;
