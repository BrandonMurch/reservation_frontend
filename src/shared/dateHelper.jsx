/* eslint-disable import/prefer-default-export */

export const getMonth = function getMonthFromNumber(monthNumber) {
  const months = [{
    long: 'January',
    Short: 'Jan',
  }, {
    long: 'February',
    Short: 'Feb',
  }, {
    long: 'March',
    Short: 'Mar',
  }, {
    long: 'April',
    Short: 'Apr',
  }, {
    long: 'May',
    Short: 'May',
  }, {
    long: 'June',
    Short: 'Jun',
  }, {
    long: 'August',
    Short: 'Aug',
  }, {
    long: 'September',
    Short: 'Sep',
  }, {
    long: 'October',
    Short: 'Oct',
  }, {
    long: 'November',
    Short: 'Nov',
  }, {
    long: 'December',
    Short: 'Dec',
  },

  ];

  return months[monthNumber];
};
