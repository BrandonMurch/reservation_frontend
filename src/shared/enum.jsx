// Thanks Andrew Nater for this trick on deep-freezing enumeration members.
const freeze = (obj) => Object.freeze(obj);

const singleValue = function deepFreezeEnumerationSingleValueMembers(...members) {
  const memberValues = {};
  members.forEach((member) => {
    memberValues[member] = freeze({ value: member.toLowerCase() });
  });

  return freeze(memberValues);
};

const keyValue = function deepFreezeEnumerationSingleValueMembers(...members) {
  const memberValues = {};
  members.forEach((member) => {
    memberValues[member.key] = freeze(member.value);
  });

  return freeze(memberValues);
};

const enumeration = {
  singleValue,
  keyValue,
};
// eslint-disable-next-line import/prefer-default-export
export { enumeration };
