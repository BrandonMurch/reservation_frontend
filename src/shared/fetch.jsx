const isStatus2xx = function checkStatusForSuccessfulResponses(status) {
  return (status >= 200 && status < 300);
};

/**
 *  Throws exception
 */
const fetchWrapper = async function fetchFromServer(path, method, body) {
  const serverAddress = 'http://localhost:8080';

  let response;
  let responseBody;
  try {
    response = await fetch(`${serverAddress}${path}`, {
      method,
      body,
      headers: { 'Content-Type': 'application/json' },
    });
    responseBody = await response.json();
  } catch (error) {
    throw new Error('Something went wrong... \n please try again later');
  }

  if (!isStatus2xx(response.status)) {
    if (responseBody && responseBody.message) {
      throw new Error(responseBody.message);
    } else {
      throw new Error('Something went wrong... \n please try again later');
    }
  }
  return responseBody;
};

export default fetchWrapper;
