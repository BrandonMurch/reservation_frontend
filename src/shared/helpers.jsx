// Thanks Andrew Nater for this trick on deep-freezing enumeration members.
const freeze = (obj) => Object.freeze(obj);

const enumeration = function deepFreezeEnumerationMembers(...members) {
  const memberValues = {};
  members.forEach((member) => {
    memberValues[member] = freeze({ value: member });
  });

  return freeze(memberValues);
};

// eslint-disable-next-line import/prefer-default-export
export { enumeration };
