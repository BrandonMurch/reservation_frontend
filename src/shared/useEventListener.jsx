import { useEffect } from 'react';

const useEventListener = (event, func) => {
  useEffect(() => {
    window.addEventListener(event, func);
    return () => {
      window.removeEventListener(event, func);
    };
  }, [event, func]);
};

export default useEventListener;
