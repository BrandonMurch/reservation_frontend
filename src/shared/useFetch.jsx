// Dependencies
import React, { useState, useEffect } from 'react';

// Components
import Loading from 'general_components/loading';
import Banner, { bannerTypes } from 'general_components/banner';

const getLoadingObject = function getLoadingComponent() {
  return {
    alternativeRender: <Loading size="large" />,
    response: null,
    error: null,
    loading: true,
  };
};

const getError = function getErrorObject(status = 500, passedInError, forceFetch) {
  let error = passedInError;
  // Server sends this message and I can't figure out where it is coming from...
  if (!error || error === 'No message available') {
    error = 'Something went wrong... \n please try again later';
  }

  return {
    forceFetch,
    status,
    error,
    alternativeRender: <Banner type={bannerTypes.ERROR} message={error} />,
    loading: false,
  };
};

const isStatus2xx = function checkStatusForSuccessfulResponses(status) {
  return (status >= 200 && status < 300);
};

const getMessageWithSubErrors = (message, subErrors) => {
  const subErrorsText = subErrors
    .map((subError) => `${subError.object} : ${subError.message}`)
    .join(', ');
    // TODO: return message \n subErrors
  return (
    <>
      <p>{message}</p>
      <p>{subErrorsText}</p>
    </>
  );
};

export const fetchWrapper = async function fetchFromServer(
  path,
  // Defaults: an object that contains an empty header object, and a 'GET' method
  { method = 'GET', headers = {}, ...fetchArguments } = {},
) {
  if (fetchArguments.skip) {
    return {
      status: null,
      alternativeRender: null,
      error: null,
      loading: false,
      body: null,
    };
  }
  const url = `http://localhost:8080${path}`;
  let response;
  headers['Content-Type'] = 'application/json';
  try {
    response = await fetch(url, { method, headers, ...fetchArguments });
  } catch (e) {
    return getError();
  }

  let forceFetch;
  if (response.headers && response.headers.has('forcible-request')) {
    headers.force = true;
    forceFetch = () => fetchWrapper(path, { method, headers, ...fetchArguments });
  }
  let responseBody;
  try {
    responseBody = await response.json();
  } catch (e) {
    if (!isStatus2xx(response.status)) {
      return getError(response.status, null, forceFetch);
    }

    return {
      status: response.status,
      alternativeRender: null,
      error: null,
      loading: false,
      body: null,
    };
  }

  if (!isStatus2xx(response.status)) {
    if (responseBody.subErrors && responseBody.subErrors.length > 0) {
      const message = getMessageWithSubErrors(response.message, response.subErrors);
      return getError(response.status, message, forceFetch);
    }
    if (responseBody.message) {
      return getError(response.status, responseBody.message, forceFetch);
    }
    return getError(response.status, null, forceFetch);
  }
  return {
    status: response.status,
    alternativeRender: null,
    response: responseBody,
    error: null,
    loading: false,
  };
};

/**
 * @returns {alternativeRender, response, error, loading}
 * alternativeRender - REACT JSX - different components, ex. error, loading
 * response - OBJECT - response from server
 * error - STRING - for custom error handling
 * loading - BOOLEAN - if the screen is still loading
 *
 * @param {*} url
 * @param {*} method
 * @param {*} body
 */
const useFetch = function useFetch(path, args, toggle = false) {
  const [fetchResponse, setFetchResponse] = useState(getLoadingObject());

  useEffect(() => {
    const getResponse = (async () => {
      setFetchResponse(await fetchWrapper(path, args));
    });
    getResponse();
  // If this is set to change with args, it just constantly reloads the page...
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [path, toggle]);

  return fetchResponse;
};

export default useFetch;
