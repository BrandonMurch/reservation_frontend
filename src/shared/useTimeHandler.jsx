import { useReducer } from 'react';
import moment from 'moment';

const useTimeHandler = (initialDate = moment()) => {
  const initialDateObject = {
    dateObject: moment(initialDate),
  };
  const reducer = ((state, action) => {
    switch (action.type) {
      case 'prev':
        return { dateObject: state.dateObject.subtract(1, action.unit) };
      case 'next':
        return { dateObject: state.dateObject.add(1, action.unit) };
      case 'current':
        return { dateObject: moment() };
      case 'goTo':
        return { dateObject: moment(action.date) };
      case 'set':
        return { dateObject: state.dateObject.set(action.unit, action.number) };
      default: throw new Error('No such action exists.');
    }
  });
  const [{ dateObject }, dispatchDate] = useReducer(reducer, initialDateObject);

  return { dateObject, dispatchDate };
};

export default useTimeHandler;
