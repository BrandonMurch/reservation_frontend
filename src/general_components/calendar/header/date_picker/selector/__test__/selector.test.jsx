// Dependencies
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import moment from 'moment';
import { getMonth } from 'shared/dateHelper';

// Components
import Selector from '../index';

describe('<Selector />', () => {
  let component;
  let container = null;
  const props = {
    dateObject: moment(),
    dispatchDate: jest.fn(),
  };

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    component = render(<Selector {...props} />);
  });

  afterEach(() => {
    props.dispatchDate.mockClear();
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('should match snapshot', () => {
    const tree = create(<Selector {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should change the year when a year button is pressed', () => {
    const year = props.dateObject.year() + 1;
    const button = component.getByRole('button', { name: year.toString() });
    fireEvent.click(button);
    expect(props.dispatchDate).toHaveBeenCalledWith({
      unit: 'years',
      number: year,
      type: 'set',
    });
  });

  it('should augment the year when the up button is pressed', () => {
    const button = component.getByLabelText('years up');
    fireEvent.click(button);
    expect(props.dispatchDate).toHaveBeenCalledWith({
      unit: 'years',
      number: -1,
      type: 'jumpUnit',
    });
  });

  it('should decrement the year when the down button is pressed', () => {
    const button = component.getByLabelText('years down');
    fireEvent.click(button);
    expect(props.dispatchDate).toHaveBeenCalledWith({
      unit: 'years',
      number: 1,
      type: 'jumpUnit',
    });
  });
  it('should change the month when a month button is pressed', () => {
    const month = props.dateObject.month() + 1;
    const monthText = getMonth(month).short;
    const button = component.getByRole('button', { name: monthText });
    fireEvent.click(button);
    expect(props.dispatchDate).toHaveBeenCalledWith({
      unit: 'months',
      number: month,
      type: 'set',
    });
  });

  it('should augment the month when the up button is pressed', () => {
    const button = component.getByLabelText('months up');
    fireEvent.click(button);
    expect(props.dispatchDate).toHaveBeenCalledWith({
      unit: 'months',
      number: -1,
      type: 'jumpUnit',
    });
  });

  it('should decrement the month when the down button is pressed', () => {
    const button = component.getByLabelText('months down');
    fireEvent.click(button);
    expect(props.dispatchDate).toHaveBeenCalledWith({
      unit: 'months',
      number: 1,
      type: 'jumpUnit',
    });
  });
});
