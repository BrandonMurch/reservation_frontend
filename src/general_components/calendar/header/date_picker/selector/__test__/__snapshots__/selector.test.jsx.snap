// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selector /> should match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="monthContainer"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <button
      aria-label="months up"
      className="directionButton"
      onClick={[Function]}
      type="button"
    >
      &lt;
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      Aug
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      Sep
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      Oct
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      Nov
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      Dec
    </button>
    <button
      aria-label="months down"
      className="directionButton"
      onClick={[Function]}
      type="button"
    >
      &gt;
    </button>
  </div>
  <div
    className="yearContainer"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <button
      aria-label="years up"
      className="directionButton"
      onClick={[Function]}
      type="button"
    >
      &lt;
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      2018
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      2019
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      2020
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      2021
    </button>
    <button
      className="wheelButton"
      onClick={[Function]}
      type="button"
    >
      2022
    </button>
    <button
      aria-label="years down"
      className="directionButton"
      onClick={[Function]}
      type="button"
    >
      &gt;
    </button>
  </div>
</div>
`;
